doc/doctorj.1
.\"Generated by db2man.xsl. Don't modify this, modify the source.
.de Sh \" Subsection
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Ip \" List item
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.TH "DOCTORJ" 1 "" "" ""
.SH NAME
doctorj \- Javadoc analyzer and spell-checker
.SH "SYNOPSIS"
.ad l
.hy 0
.HP 8
\fBdoctorj\fR [\fB\-\-emacs\fR] [\fB\-\-level\fR\fI=LEVEL\fR] [\fB\-\-warning\fR\fI=LEVEL\fR] [\fB\-\-tabwidth\fR\fI=NUM\fR] [\fB\-\-dictionaries\fR\fI=LIST\fR] [\fB\-\-source\fR\fI=VERSION\fR] [\fB\-\-verbose\fR] {\fIFILE\&.\&.\&.\fR}
.ad
.hy

.SH "DESCRIPTION"

.PP
doctorj is an application to analyze Javadoc comments against Java code, verifying its integrity and format\&. It also spell\-checks Javadoc comments\&.

.SH "OPTIONS"

.TP
\fB\-\-emacs\fR
Whether to list violations in Emacs form, i\&.e\&., a single line of the form:

8:48:9:4: An error occurred on this line\&. 

By default, violations are displayed in ``long form'', showing the context\&.

.TP
\fB\-\-level\fR\fI=LEVEL\fR
Sets the warning/error level\&. A level of 0 will result in only errors reported\&. A level of 1 means that Javadoc for public and abstract items (classes, interfaces, methods, constructors, and fields) will be checked\&. A level of 2 will result in the addition of checking of protected\-access items; 3 is for package\-access items, and 4 will check all items\&. The default is 4\&.

.TP
\fB\-\-warning\fR\fI=LEVEL\fR
This is the same as \-\-level; it exists for backward compatibility, and is deprecated\&.

.TP
\fB\-\-tabwidth\fR\fI=NUM\fR
Sets the number of spaces to use for alignment with tabs\&. By default, this value is 4\&.

.TP
\fB\-\-dictionaries\fR\fI=LIST\fR
Sets the dictionaries (word lists) to use for spell\-checking\&. \fILIST\fR is a comma\-delimited string of paths to the word list files\&. By default, doctorj uses the word list at \fI/usr/share/doctorj/words\&.LOCALE\fR, where \fILOCALE\fR is of the standard Java form\&. doctorj is distributed with three word lists: words\&.en_CA (Canadian English), words\&.en_UK (United Kingdom English), and words\&.en_US (United States English)\&.

.TP
\fB\-\-verbose\fR
Produces debugging output\&. Note that this signficantly impedes performance\&.

.TP
\fB\-\-source\fR\fI=VERSION\fR
Sets the source version, which is used to determine the type of grammar with which to parse the input files\&. \fIVERSION\fR may be 1\&.3, 1\&.4, or 1\&.5\&. The default is 1\&.4\&.

.TP
\fBFILE\fR
Specifies the files to check\&. If the argument specified is a directory, all files ending in ``\&.java'' under that directory will be analyzed\&.

.SH "EXAMPLES"

.PP
To check a single file: 

.IP
% doctorj String\&.java

.PP
To check multiple files: 

.IP
% doctorj String\&.java StringBuffer\&.java Object\&.java

.PP
To check all files below a directory: 

.IP
% doctorj /proj/app/src

.PP
To run at the maximum warning level: 

.IP
% doctorj \-\-warning=4 \&.

.PP
To specify a different dictionary: 

.IP
% doctorj \-\-dictionaries=/usr/share/dict/words \&.

.SH "CONFIGURATION"

.PP
\fB$HOME/\&.doctorjrc\fR

.PP
The run\-control file, of the form: 

.IP

    emacs: true
    warning: 4
    tabwidth: 8
    dictionaries: /usr/share/dict/words,/home/me/etc/terms
    verbose: false
          

.PP
Dictionary (word list) files contain one word per line, for example:

.PP
 

.IP

    adaxial
    cepaceous
    sabaton
    vespiaries
    eujifferous
         

.SH "AUTHOR"

.PP
Jeff Pace<jpace@incava\&.org>

.PP
The DoctorJ web site is at \fIhttp://www.incava.org/projects/java/doctorj\fR\&.

.SH "BUGS"

.PP
Bugs and issues should be reported to the author: Jeff Pace<jpace@incava\&.org>  http://www\&.sf\&.net/projects/doctorj: \fIhttp://www.sf.net/projects/doctorj\fR 

.SH "SEE ALSO"

.PP
 \fBjava\fR(1), \fBjavadoc\fR(1) 

.SH "COPYRIGHT"

.PP
Copyright (c) 2002\-2006, Jeff Pace\&. All Rights Reserved\&. This module is free software\&. It may be used, redistributed and/or modified under the terms of the GNU Lesser General Public License: \fIhttp://www.gnu.org/licenses/lgpl.html\fR

.PP
Copyright (c) 2003, InfoEther, LLC\&. All rights reserved\&. This product includes software developed in part by support from the Defense Advanced Research Project Agency (DARPA)\&.



src/main/java/org/incava/doctorj/DoctorJ.java
package org.incava.doctorj;

import java.io.*;
import java.util.*;
import net.sourceforge.pmd.ast.*;
import org.incava.analysis.*;
import org.incava.ijdk.io.Find;
import org.incava.pmdx.SimpleNodeUtil;
import org.incava.util.TimedEvent;
import org.incava.util.TimedEventSet;


public class DoctorJ {

    private TimedEventSet _totalInit = new TimedEventSet();

    private TimedEventSet _totalParse = new TimedEventSet();

    private TimedEventSet _totalAnalysis = new TimedEventSet();

    private JavaParser _parser = null;

    private final Report _report;

    private final JavaParserVisitor _analyzer;

    private int _exitValue;

    private int _nFiles;

    public DoctorJ(String[] args) {
        tr.Ace.set(true, 25, 4, 20, 25);
        tr.Ace.setOutput(tr.Ace.VERBOSE, tr.Ace.LEVEL4);
        tr.Ace.setOutput(tr.Ace.QUIET,   tr.Ace.LEVEL2);

        _exitValue = 0;
        _nFiles = 0;

        Options  opts = Options.get();
        String[] names = opts.process(args);

        tr.Ace.log("args", args);
        tr.Ace.log("names", names);
        tr.Ace.log("properties", System.getProperties());

        if (opts.emacsOutput) {
            _report = new TerseReport(System.out);
        }
        else {
            _report = new ContextReport(System.out);
        }
        
        _analyzer = new JavadocAnalyzer(_report);

        for (int ni = 0; ni < names.length; ++ni) {
            String name = names[ni];
            process(name);
        }
        
        if (_nFiles > 1) {
            long total = _totalInit.duration + _totalParse.duration + _totalAnalysis.duration;
            tr.Ace.log("total time: " + total);
            tr.Ace.log("init      : " + _totalInit.duration);
            tr.Ace.log("parse     : " + _totalParse.duration);
            tr.Ace.log("analysis  : " + _totalAnalysis.duration);
            tr.Ace.log("#files    : " + _nFiles);
        }
    }

    public int getExitValue() {
        return _exitValue;
    }

    protected void process(String name) {
        File fd = new File(name);
        if (fd.exists()) {
            if (fd.isDirectory()) {
                tr.Ace.log("processing directory");
                String[] contents = fd.list();
                Arrays.sort(contents);
                for (int ci = 0; ci < contents.length; ++ci) {
                    String nm = contents[ci];
                    String fullName = name + File.separator + nm;
                    File   f = new File(fullName);
                    if (f.isDirectory()) {
                        process(fullName);
                    }
                    else if (f.isFile() && nm.endsWith(".java")) {
                        processFile(fullName);
                    }
                    else {
                        tr.Ace.log("not a match", f);
                    }   
                }
            }
            else if (fd.isFile()) {
                processFile(name);
            }
        }
        else {
            System.err.println("File " + name + " not found.");
        }
    }
    
    protected void processFile(String fileName) {
        ++_nFiles;

        if (initParser(fileName)) {
            ASTCompilationUnit cu = parse(fileName);
            if (cu != null) {
                analyze(cu);
            }
        }
    }

    protected boolean initParser(String fileName) {
        tr.Ace.log("fileName", fileName);

        TimedEvent init = new TimedEvent(_totalInit);
        try {
            _report.reset(new File(fileName));
                
            // = fromIsStdin ? new FileReader(FileDescriptor.in) :
            FileReader     rdr = new FileReader(fileName);
            JavaCharStream jcs = new JavaCharStream(rdr);
            
            _parser = new JavaParser(jcs);

            String src = Options.get().source;
            if (src.equals("1.3")) {
                tr.Ace.log("setting as 1.3");
                _parser.setJDK13();
            }
            else if (src.equals("1.4")) {
                tr.Ace.log("leaving as 1.4");
            }
            else if (src.equals("1.5")) {
                tr.Ace.log("setting as 1.5");
                _parser.setJDK15();
            }
            else {
                System.err.println("ERROR: source version '" + src + "' not recognized");
                System.exit(-1);
            }
            init.end();

            tr.Ace.log("init: " + init.duration);

            return true;
        }
        catch (FileNotFoundException e) {
            System.out.println("File " + fileName + " not found.");
            _exitValue = 1;

            return false;
        }
        catch (IOException e) {
            System.out.println("Error opening " + fileName + ": " + e);
            _exitValue = 1;

            return false;
        }
        catch (TokenMgrError tme) {
            System.out.println("Error parsing (tokenizing) " + fileName + ": " + tme.getMessage());
            _exitValue = 1;

            return false;
        }
    }

    protected ASTCompilationUnit parse(String fileName) {
        tr.Ace.log("running parser");
            
        try {
            TimedEvent         parse = new TimedEvent(_totalParse);
            ASTCompilationUnit cu = _parser.CompilationUnit();
            parse.end();

            tr.Ace.log("parse: " + parse.duration);

            return cu;
        }
        catch (ParseException e) {
            System.out.println("Parse error in " + fileName + ": " + e.getMessage());
            _exitValue = 1;

            return null;
        }
    }

    protected void analyze(ASTCompilationUnit cu) {
        TimedEvent analysis = new TimedEvent(_totalAnalysis);
        cu.jjtAccept(_analyzer, null);

        _report.flush();
        analysis.end();

        tr.Ace.log("analysis: " + analysis.duration);
    }

    public static void main(String[] args) {
        tr.Ace.setVerbose(false);
        
        DoctorJ drj = new DoctorJ(args);
        System.exit(drj.getExitValue());
    }

}
